app/src/main/java/io/appium/uiautomator2/model/XPathFinder.java

```diff
@@ -51,235 +51,197 @@
-      NodeInfoList list = new NodeInfoList();
+    private XPathFinder(String xPathString) {
+        this.xPathString = Preconditions.checkNotNull(xPathString);
+    }
 
-      int nodesLength = nodes.getLength();
-        for (int i = 0; i < nodesLength; i++) {
-          if (nodes.item(i).getNodeType() == Node.ELEMENT_NODE && !FROM_DOM_MAP.get(nodes.item(i)).getClassName().equals("hierarchy")) {
-              list.addToList(FROM_DOM_MAP.get(nodes.item(i)).node);
-          }
+    @Override
+    public NodeInfoList find(UiElement context) {
+        final Document document;
+        try {
+            document = DocumentBuilderFactory.newInstance()
+                    .newDocumentBuilder()
+                    .newDocument();
+        } catch (ParserConfigurationException e) {
+            throw new UiAutomator2Exception(e);
+        }
+        final Element domNode = toDOMElement((UiElement<?, ?>) context, document);
+        document.appendChild(domNode);
+        try {
+            final NodeList nodes = (NodeList) XPATH_COMPILER
+                    .compile(xPathString)
+                    .evaluate(domNode, XPathConstants.NODESET);
+            final NodeInfoList matchesList = new NodeInfoList();
+            for (int i = 0; i < nodes.getLength(); i++) {
+                if (nodes.item(i).getNodeType() != Node.ELEMENT_NODE) {
+                    continue;
+                }
+
+                final NamedNodeMap attributes = nodes.item(i).getAttributes();
+                final Node uuidAttribute = attributes.getNamedItem(UUID_ATTRIBUTE);
+                if (uuidAttribute == null) {
+                    continue;
+                }
+                if (!UI_ELEMENTS_MAP.containsKey(uuidAttribute.getNodeValue()) ||
+                        UI_ELEMENTS_MAP.get(uuidAttribute.getNodeValue()).getClassName().equals("hierarchy")) {
+                    continue;
+                }
+
+                matchesList.addToList(UI_ELEMENTS_MAP.get(uuidAttribute.getNodeValue()).node);
+            }
+            return matchesList;
+        } catch (XPathExpressionException e) {
+            throw new ElementNotFoundException(e);
         }
-      return list;
-    } catch (XPathExpressionException e) {
-      throw new ElementNotFoundException( e);
-    } finally {
-      try {
-        getDocument().removeChild(domNode);
-      } catch (DOMException e) {
-        Logger.error(e, "Failed to clear document");
-        document = null; // getDocument will create new
-      }
     }
-  }
 
-  public static NodeInfoList getNodesList(String xpathExpression,  AccessibilityNodeInfo nodeInfo) throws InvalidSelectorException, ParserConfigurationException, UiAutomator2Exception {
-    if(nodeInfo == null) {
-      XPathFinder.refreshUiElementTree();
-    } else {
-      XPathFinder.refreshUiElementTree(nodeInfo);
+    public static NodeInfoList getNodesList(String xpathExpression,
```

Comment:
It seems that static modifiers are unnecessary here and below.

Reply:
with static is better from performance perspective than without, since such method calls are not going through vtable.

Reply:
It's not true nowadays.
https://ideone.com/goZurZ

Reply:
yep, the difference is very tiny. However, it still helps me to easily distinguish "utility" methods from these, that use class properties.